pkg/init/cmd/service/logging.go

```diff
@@ -0,0 +1,177 @@
+package main
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"path/filepath"
+	"syscall"
+
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	errLoggingNotEnabled = errors.New("logging system not enabled")
+	logWriteSocket       = "/var/run/external-logging.sock"
+	logReadSocket        = "/var/run/memlogdq.sock"
+)
+
+const (
+	logDumpCommand byte = iota
+)
+
+// Log sets up a log destination for a named service.
+type Log interface {
+	Path(name string) string            // Path of the log file (or FIFO)
+	Open(name string) (*os.File, error) // Opens a log file directly
+	DumpAll()                           // Copies all the logs to the console
+}
+
+// GetLog returns the log destination we should use.
+func GetLog(logDir string) Log {
+	// is an external logging system enabled?
+	_, err := os.Stat(logWriteSocket)
+	if !os.IsNotExist(err) {
+		return &remoteLog{
+			fifoDir: "/var/run",
+		}
+	}
+	return &fileLog{
+		dir: logDir,
+	}
+}
+
+type fileLog struct {
+	dir string
+}
+
+// Path returns the name of a log file path for the named service.
+func (f *fileLog) Path(name string) string {
+	// We just need this to exist. If we cannot write to the directory,
+	// we'll discard output instead.
+	file, err := f.Open(name)
+	if err != nil {
+		return "/dev/null"
+	}
+	_ = file.Close()
+	return filepath.Join(f.dir, name+".log")
```

Comment:
I'm confused, `file` had a different name (and potentially path, depending on `cwd`) to what is being returned.

Reply:
This just confused me as well (!) It's confusing because `f.Open(name)` looks like a call to `os.Open(path)` but is actually a call to
```golang
func (f *fileLog) Open(name string) (*os.File, error) {
	return os.Create(filepath.Join(f.dir, name+".log"))
}
```
so I think it is working by computing the path twice. I'll see if I can rename or otherwise refactor `Open` (and `Path`) to avoid them being confused with `os.Open` etc. Since this behaviour was copied from the previous code I re-checked what happens if you omit it -- unfortunately `containerd` fails to start the task if the path doesn't exist. I'll mention that in the comments.

Reply:
`*os.File` does have a `Name` method but a) I'm not convinced that'll be the full path b) I think the return should be an `os.WriteCloser` anyway.

Maybe just have `fileLog.Open` return the name as well as the `os.WriteCloser` (and the err)?

Reply:
Or since the `fileLog.Open` is also part of the public interface (meaning you'd have to change the other one too, where it might make sense) make an internal helper which `Open` and `Path` both call internally.

Reply:
Sorry I just raced with your comment and updated with something similar-ish. I've made a common function
```golang
func (f *fileLog) localPath(n string) string {
	return filepath.Join(f.dir, n+".log")
}
```
and now `Path` and `Open` call this function and then directly call `os.*` functions (rather than one calling the other which was hard to read):
```golang
// Path returns the name of a log file path for the named service.
func (f *fileLog) Path(n string) string {
	path := f.localPath(n)
	// We just need this to exist, otherwise containerd will say:
	//
	//   ERRO[0000] failed to create task error="failed to start io pipe
	//     copy: containerd-shim: opening /var/log/... failed: open
	//     /var/log/...: no such file or directory: unknown"
	file, err := os.Create(path)
	if err != nil {
		// If we cannot write to the directory, we'll discard output instead.
		return "/dev/null"
	}
	_ = file.Close()
	return path
}

// Open a log file for the named service.
func (f *fileLog) Open(n string) (io.WriteCloser, error) {
	return os.OpenFile(f.localPath(n), os.O_WRONLY|os.O_CREATE, 0644)
}
```
-- does that seem reasonable?